internal/router/router.go

```diff
@@ -0,0 +1,95 @@
+package router
+
+import (
+	"fmt"
+	"github.com/fabric8-services/fabric8-jenkins-idler/internal/api"
+	"github.com/julienschmidt/httprouter"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"net/http"
+	"time"
+)
+
+var routerLogger = log.WithFields(log.Fields{"component": "router"})
+
+const (
+	defaultHttpServerPort = 8080
+	shutdownTimeout       = 5
+)
+
+// Router implements an HTTP server, exposing the REST API of the Idler
+type Router struct {
+	port int
+	srv  *http.Server
+}
+
+// NewRouter creates a new HTTP router for the Idler on the default port.
+func NewRouter(api api.IdlerAPI) *Router {
+	router := createRouter(api)
+	srv := &http.Server{
+		Addr:    fmt.Sprintf(":%d", defaultHttpServerPort),
+		Handler: router,
+	}
+
+	return &Router{port: defaultHttpServerPort, srv: srv}
+}
+
+// NewRouter creates a new HTTP router for the Idler on the specified port.
+func NewRouterWithPort(api api.IdlerAPI, port int) *Router {
+	router := createRouter(api)
+	srv := &http.Server{
+		Addr:    fmt.Sprintf(":%d", port),
+		Handler: router,
+	}
+
+	return &Router{port: port, srv: srv}
+}
+
+// Start starts the HTTP router
+func (r *Router) Start(done <-chan interface{}) <-chan interface{} {
+	terminated := make(chan interface{})
+	go func() {
+		go func() {
+			defer routerLogger.Infof("Idler router stopped.")
+			defer close(terminated)
```

Comment:
Just to confirm I understand this correctly - before the `go func` returns, `terminated` is closed, which unblocks https://github.com/fabric8-services/fabric8-jenkins-idler/pull/88/files#diff-dd81eea76ff6680061606f2bdbaf421aR125 - correct?

Reply:
Right. Works quite well for a single goroutine. For multiple, you need to wait on multiple channels or use a WaitGroup. Most likely I will change this when I change the code for watching the build and deployment configs.